pkg/agent/util/iptables/iptables.go

```diff
@@ -145,18 +150,25 @@ func (c *Client) EnsureRule(table string, chain string, ruleSpec []string) error
 		if err := ipt.Append(table, chain, ruleSpec...); err != nil {
 			return fmt.Errorf("error appending rule %v to table %s chain %s: %v", ruleSpec, table, chain, err)
 		}
+		klog.V(2).InfoS("Appended a rule", "rule", "table", "chain", "protocol", ruleSpec, table, chain, p.String())
+		return nil
 	}
-	klog.V(2).Infof("Appended rule %v to table %s chain %s", ruleSpec, table, chain)
-	return nil
+	if protocol == ProtocolDual {
+		if err := appendRule(ProtocolIPv4, table, chain, ruleSpec); err != nil {
+			return err
+		}
+		if err := appendRule(ProtocolIPv6, table, chain, ruleSpec); err != nil {
+			return err
+		}
+		return nil
+	}
+	return appendRule(protocol, table, chain, ruleSpec)
 }
 
-// InsertRule checks if target rule already exists, inserts it if not.
+// InsertRule checks if target rule already exists with the protocol, inserts it if not.
 func (c *Client) InsertRule(protocol Protocol, table string, chain string, ruleSpec []string) error {
-	for idx := range c.ipts {
-		ipt := c.ipts[idx]
-		if !matchProtocol(ipt, protocol) {
-			continue
-		}
+	insertRule := func(p Protocol, table string, chain string, ruleSpec []string) error {
```

Comment:
My intention was to use a map to fasten the process for single stack case. However, since it is just a loop of 2 elements, it doesn't make much difference.